ledger-api/grpc-definitions/com/daml/ledger/api/v2/commands.proto

```diff
@@ -0,0 +1,195 @@
+  // Optional
+  repeated string read_as = 11;
+
+  // A unique identifier to distinguish completions for different submissions with the same change ID.
+  // Typically a random UUID. Applications are expected to use a different UUID for each retry of a submission
+  // with the same change ID.
+  // Must be a valid LedgerString (as described in ``value.proto``).
+  //
+  // If omitted, the participant or the committer may set a value of their choice.
+  // Optional
+  string submission_id = 12;
+
+  // Additional contracts used to resolve contract & contract key lookups.
+  // Optional
+  repeated com.daml.ledger.api.v1.DisclosedContract disclosed_contracts = 13;
+
+  // Must be a valid domain ID
+  // Required
+  string domain_id = 14;
+}
+
+message ReassignmentCommand {
+  // Identifier of the on-ledger workflow that this command is a part of.
+  // Must be a valid LedgerString (as described in ``value.proto``).
+  // Optional
+  string workflow_id = 1;
+
+  // Uniquely identifies the application or participant user that issued the command.
+  // Must be a valid ApplicationIdString (as described in ``value.proto``).
+  // Required unless authentication is used with a user token or a custom token specifying an application-id.
+  // In that case, the token's user-id, respectively application-id, will be used for the request's application_id.
+  string application_id = 2;
+
+  // Uniquely identifies the command.
+  // The triple (application_id, submitter, command_id) constitutes the change ID for the intended ledger change.
+  // The change ID can be used for matching the intended ledger changes with all their completions.
+  // Must be a valid LedgerString (as described in ``value.proto``).
+  // Required
+  string command_id = 3;
+
+  // Party on whose behalf the command should be executed.
+  // If ledger API authorization is enabled, then the authorization metadata must authorize the sender of the request
+  // to act on behalf of the given party.
+  // Must be a valid PartyIdString (as described in ``value.proto``).
+  // Required
+  string submitter = 4;
+
+  oneof command {
+    UnassignCommand unassign_command = 5;
+    AssignCommand assign_command = 6;
+  }
+
+  // A unique identifier to distinguish completions for different submissions with the same change ID.
+  // Typically a random UUID. Applications are expected to use a different UUID for each retry of a submission
+  // with the same change ID.
+  // Must be a valid LedgerString (as described in ``value.proto``).
+  //
+  // If omitted, the participant or the committer may set a value of their choice.
+  // Optional
+  string submission_id = 7;
+}
+
+// Assign a contract
+message UnassignCommand {
+  // The ID of the contract the client wants to unassign.
+  // Must be a valid LedgerString (as described in ``value.proto``).
+  // Required
+  string contract_id = 1;
+
+  // The ID of the source domain
+  // Must be a valid domain ID
+  // Optional
```

Comment:
In theory, this can be optional and only used for validation. Currently, we validate that it is populated and we use that to grab to SyncDomain on which we submit the transfer-out. I'd make it required for now and relax in the future if we can.

Reply:
Sure, I like it since it makes it more explicit about what state we are talking about.
To make it more explicit we could add also the transfer counter as well, WDYT?

Reply:
making it required